backend/funix/widget/__init__.py

```diff
@@ -1,76 +1,172 @@
-import json
+"""
+This file contains the widget system for the funix project.
+"""
+
+from json import dumps
 from typing import TypedDict
 
+# Widget update
+widgets_update_config = {}
 
-def generate_widget_config(widget: str, config: TypedDict) -> tuple[str, TypedDict]:
-    return widget, config
 
+def register_widget_update_config(widget: str) -> callable:
+    """
+    Register a widget update config function.
+    You should register a function, that takes *args, and return a dict.
+    The dict will be merged with the original config dict.
+    This is used to update the config of a widget and solve the problem of the config of a widget is not complete.
 
-widgets_update_config = {}
+    Parameters:
+        widget(str): The name of the widget.
 
+    Returns:
+        The decorator.
+    """
+    def decorator(func: callable) -> None:
+        """
+        Decorator for register a widget update config function.
 
-def register_widget_update_config(widget: str):
-    def decorator(func):
+        Parameters:
+            func(callable): The function.
+        """
         widgets_update_config[widget] = func
 
     return decorator
 
 
-def dump_frontend_config(config: tuple[str, TypedDict]) -> str:
+def dump_frontend_config(config: (str, TypedDict)) -> str:
+    """
+    Convert the config to a string that can be used in frontend. Will auto update the config if this widget has an
+    update config function.
+    And this is the expanded list_config definition 🤣:
+
+    ```python
+    list_config = json.dumps(
+        list(
+            widgets_update_config[
+                config[0]
+            ]  # Get the function in `widgets_update_config` by the widget name.
+            (
+                list(
+                    config[1].values()
+                )  # This is the sorted config list
+            ).values()  # get the values of the sorted merged config list
+        )
+    )
+    ```
+
+    Parameters:
+        config((str, TypedDict)): The config.
+            str: Widget name.
+            TypedDict: Widget Config.
+    """
     if config[0] in widgets_update_config:
-        list_config = json.dumps(list(widgets_update_config[config[0]](list(config[1].values())).values()))
+        list_config = dumps(list(widgets_update_config[config[0]](list(config[1].values())).values()))
         return f"{config[0]}{list_config}"
-    return f"{config[0]}{json.dumps(list(config[1].values()))}"
+    return f"{config[0]}{dumps(list(config[1].values()))}"
```

Comment:
can you help me understand why we only do it for config[0]?

Reply:
Oh I see, it's a tuple....
Consider later use a `dataclass` object, so you can do `config.name` and `config.value` to reference them, as opposed to `config[0]`, and `config[1]`.

Reply:
I think this is okay: `def dump_frontend_config(widget_name: str, config: TypedDict) -> str:` (before `def dump_frontend_config(config: (str, TypedDict)) -> str:`)